tket/src/Converters/ChoiMixTableauConverters.cpp

```diff
@@ -30,641 +29,732 @@ ChoiMixTableau circuit_to_cm_tableau(const Circuit& circ) {
   }
+  // Postselect rows
+  for (unsigned r = 0; r < n_postselected; ++r) {
+    unsigned final_row = tab.get_n_rows() - 1;
+    ChoiMixTableau::row_tensor_t row = tab.get_row(final_row);
+    if (row.second.string.map.size() != 0 || row.first.string.map.size() != 1 ||
+        row.first.string.map.begin()->second != Pauli::Z)
+      throw std::logic_error(
+          "Unexpected error during post-selection identification in "
+          "ChoiMixTableau synthesis");
+    Qubit post_selected_qb = row.first.string.map.begin()->first;
+    // Multiply other rows to remove Z_qb components
+    unsigned qb_col = tab.col_index_.left.at(ChoiMixTableau::col_key_t{
+        post_selected_qb, ChoiMixTableau::TableauSegment::Input});
+    for (unsigned s = 0; s < final_row; ++s)
+      if (tab.tab_.zmat_(s, qb_col)) tab.tab_.row_mult(final_row, s);
+    // Post-select on correct phase
+    if (row.second.coeff == -1.)
+      in_circ.add_op<Qubit>(OpType::X, {post_selected_qb});
+    tab.remove_row(final_row);
+    post_selected.insert(post_selected_qb);
+    tab.discard_qubit(post_selected_qb, ChoiMixTableau::TableauSegment::Input);
+  }
+}
 
-  // Fix phases of zero_initialised qubits
-  std::set<Qubit> zero_initialised;
-  Circuit out_circ(output_qubits, {});
-  for (unsigned r = 0; r < tab.get_n_rows(); ++r) {
+void ChoiMixBuilder::solve_initialised_subspace() {
+  // Input-first gaussian elimination now sorts the remaining rows into the
+  // collapsed subspace followed by the zero-initialised subspace and the
+  // collapsed subspace rows are in upper echelon form over the inputs, giving
+  // unique leading columns and allowing us to solve them with CXs by
+  // column-wise gaussian elimination; same for zero-initialised rows over the
+  // outputs
+  tab.canonical_column_order(ChoiMixTableau::TableauSegment::Input);
+  tab.gaussian_form();
+
+  // Reduce the zero-initialised space to a minimal set of qubits using CX gates
+  unsigned n_collapsed = 0;
+  for (; n_collapsed < tab.get_n_rows(); ++n_collapsed) {
+    if (tab.get_row(n_collapsed).first.string.map.empty()) break;
+  }
+  unsigned n_ins = tab.get_n_inputs();
+  unsigned n_outs = tab.get_n_outputs();
+  MatrixXb subtableau =
+      tab.tab_.zmat_.bottomRightCorner(tab.get_n_rows() - n_collapsed, n_outs);
+  std::vector<std::pair<unsigned, unsigned>> col_ops =
+      leading_column_gaussian_col_ops(subtableau);
+  for (const std::pair<unsigned, unsigned>& op : col_ops) {
+    unsigned tab_ctrl_col = n_ins + op.second;
+    unsigned tab_trgt_col = n_ins + op.first;
+    tab.tab_.apply_CX(tab_ctrl_col, tab_trgt_col);
+    ChoiMixTableau::col_key_t ctrl = tab.col_index_.right.at(tab_ctrl_col);
+    ChoiMixTableau::col_key_t trgt = tab.col_index_.right.at(tab_trgt_col);
+    out_circ_tp.add_op<Qubit>(OpType::CX, {ctrl.first, trgt.first});
+  }
+  // Initialise rows
+  for (unsigned r = tab.get_n_rows(); r > n_collapsed;) {
+    --r;  // This always refers to the final row in the tableau
```

Comment:
```suggestion
  for (unsigned r = tab.get_n_rows(); r-- > n_collapsed;) {
    // r always refers to the final row in the tableau
```
Maybe slightly nicer like this?

Reply:
Personal taste. I prefer to not use increment/decrement in the middle of expressions where possible as I find it less clear and have to mentally hold onto two different values when reading a single line of code and worry about pre/postfix confusion. Certainly a good compiler should be comfortable with equating the two.

I guess there is the perspective in this case that the loop variable has the intended semantics throughout the entire loop body and all iteration logic is kept to the header, but this does come at the cost of a more complicated header

Reply:
Yeah that's exactly why I prefer it (keeping the logic in the header). But happy to keep as is.